### ISSUE TITLE
[iCloud] account thing won't login after openhab restart

### ISSUE THREAD

**Issue by vsubbotskyy** [openhab/openhab-addons#17590]
## Expected Behavior
Should login to iCloud 
## Current Behavior
Right now when I save the binding it won't login and won't trigger 2fa

## Steps to Reproduce (for Bugs)

Restart openhab service, observe iCloud account thing

## Context
This started recently I guess. I found only one post on community portal, and it is very recent:
https://community.openhab.org/t/icloud-binding-authentication-problem/159739

## Your Environment
- openHAB 4.2.1 - Release Build
- iCloud Binding 4.2.1

**Comment by phuongpham** [comment]
Check the fix for pyicloud at https://github.com/picklepete/pyicloud/issues/456

for base.py

**Comment by maihacke** [comment]
I had a look into this. The problem is that Apple changed there authentication mechanism. They know use SRP, which is not supported by the binding.
Having this analyzed a bit it should be possible to adopt the binding accordingly.

Currently there is one challenge with has to be discussed.
We will need an SRP implementation. There exists several options, e.g. Bouncycastle or Nimbus SRP:
Since BC is already included in OH I did few tests with BC, but soon was blocked by the following problem:

OH includes pcpkix-jdk18on-1.77.jar which is repackaged in org.openhab.io.jetty.certificate.jar. BCs SRP implementation is part of pcprov-jdkon-1.77.jar. When adding this to the classpath I get an security exceptions because the jar signatures do not match. I saw some older discussions around this in the forum (e.g. https://community.openhab.org/t/io-jetty-certificate-embeds-bouncy-castle-crypto-classes-resulting-in-classloader-error/153020/1).
From what I see currently it would not be possible to use SRP from BC without changes to certificate.jar form OH-core.
So it might be easier to directly go to Nimbus.

I'm not sure who is responsible for these type of questions, so any option on that, @kaikreuzer ?

**Comment by lochmueller** [comment]
Short note: Same problem on my machine with openHAB 5.0.0.M2

**Comment by maihacke** [comment]
I started some experiments with the new authentication scheme, but the original problem disappeared. So it looks like Apple has reverted the changes. This was also observed by the pyicloud "people" (see comments in https://github.com/picklepete/pyicloud/issues/456).
So currently I see no issues with openhab icloud authentication. @lochmueller if you observed problemens recently it might be something else or user specific. Could you share logs here please?

**Comment by lochmueller** [comment]
Hey @maihacke 
the account was blocked by apple today again. Perhaps apple classified the openhab login process as "security relevant" and disable the icloud account :-/ I will check this in the next days...
Regads,
Tim

**Comment by mike-bike** [comment]
I have experienced the same issue with iCloud binding today. It was running stable for quite a while with 4.3x and up to now 5.0.2. At around 20:25 CET the iCloud thing went offline with error: 

`2025-11-12 20:25:33.405 [WARN ] [l.handler.ICloudAccountBridgeHandler] - iCloud authentication failed. Invalid credentials.`

Any attempt to re-enable the thing resulted in the same error message. No ask for 2F as previously.
The configured id and password are accepted via browser login to iCloud.com and the account does not seem to locked as @lochmueller observed (How could I verify this?)

Any additional info I could provide?

Thanks,
Michael

**Comment by maihacke** [comment]
I never observed this on my side. On my side it logs out from time to time, but pausing/starting the thing always resolved the issue.
To analyze this further we need some DEBUG logs.
So please enable DEBUG forborg.binding.icloud and provide the logs. Please remove security related info before posting them here.

**Comment by mike-bike** [comment]
Hi @maihacke, I used to do the same whenever the binding lost connection every couple of weeks without issues. From time to time I have then been asked for 2F, but reconnect was working fine. Since yesterday, this does no longer work. 
I had also restarted openHAB with cache cleared; same result. I am not getting asked for 2F or getting any notification on my trusted iCloud devices.
Please note, that logon to www.icloud.com with the same credentials is working as before.

Please find below corresponding iCloud DEBUG log:

```2025-11-13 13:20:43.939 [DEBUG] [l.handler.ICloudAccountBridgeHandler] - iCloud bridge handler initializing ...
2025-11-13 13:20:43.944 [DEBUG] [l.handler.ICloudAccountBridgeHandler] - iCloud bridge handler initialized.
2025-11-13 13:20:43.961 [DEBUG] [l.handler.ICloudAccountBridgeHandler] - Starting iCloud authentication (AuthState=INITIAL, Thing=icloud:account:879ff9a943)...
2025-11-13 13:20:43.962 [DEBUG] [inding.icloud.internal.ICloudService] - Checking session token validity
2025-11-13 13:20:43.967 [DEBUG] [inding.icloud.internal.ICloudSession] - iCloud request POST https://setup.icloud.com/setup/ws/1/validate.
2025-11-13 13:20:44.358 [DEBUG] [inding.icloud.internal.ICloudService] - Token is not valid. Attemping new login.
org.openhab.binding.icloud.internal.ICloudApiResponseException: Request https://setup.icloud.com/setup/ws/1/validate failed with 421.
        at org.openhab.binding.icloud.internal.ICloudSession.request(ICloudSession.java:148) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudSession.post(ICloudSession.java:98) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudService.validateToken(ICloudService.java:190) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudService.authenticate(ICloudService.java:94) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.checkLogin(ICloudAccountBridgeHandler.java:338) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.callApiWithRetryAndExceptionHandling(ICloudAccountBridgeHandler.java:164) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:127) ~[?:?]
        at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
        at java.lang.Thread.run(Unknown Source) [?:?]
2025-11-13 13:20:44.361 [DEBUG] [inding.icloud.internal.ICloudService] - Authenticating as myapple.id@mail.com...
2025-11-13 13:20:44.362 [DEBUG] [inding.icloud.internal.ICloudSession] - iCloud request POST https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true.
2025-11-13 13:20:44.826 [WARN ] [l.handler.ICloudAccountBridgeHandler] - iCloud authentication failed. Invalid credentials.
```

Setting binding to TRACE I found below response at the end of the process. Seems to be an issue at Apple which is not correctly interpreted by the binding. Looks like temporary outage of the service rather than authentication error. Might be sensible to change the binding to quote the response into the message, rather than assuming invalid credentials. 
Not sure, if that helps.

```
------

2025-11-13 13:31:05.425 [TRACE] [inding.icloud.internal.ICloudSession] - Result https://idmsa.apple.com/appleauth/auth/signin?isReme
mberMeEnabled=true 503
Headers -----
java.net.http.HttpHeaders@ab290661 { {connection=[keep-alive], content-length=[190], content-type=[text/html], date=[Thu, 13 Nov 202
5 12:31:05 GMT], server=[Apple]} }
Body -----
<html>
<head><title>503 Service Temporarily Unavailable</title></head>
<body>
<center><h1>503 Service Temporarily Unavailable</h1></center>
<hr><center>Apple</center>
</body>
</html>

------
2025-11-13 13:31:05.426 [WARN ] [l.handler.ICloudAccountBridgeHandler] - iCloud authentication failed. Invalid credentials.
```

**Comment by lochmueller** [comment]
> [@lochmueller](https://github.com/lochmueller) observed (How could I verify this?)

@mike-bike You get a warning if you login into icloud.com.

In my openHAB instance, the binding also works very instable. But I have to check this later (no high priority for me at the moment)

**Comment by cgatnet** [comment]
I have the same issue like mike-bike reported, now maybe since two days, i am no longer able to use the binding: Invalid credentials. I changed nothing, it just happened. Also using actual OH5.0.2
Also checked my iCloud login via browser. There was no problem with that.
Never experienced a blocked account by now.

And looks like it is same log at my side:

```
2025-11-13 22:50:34.340 [DEBUG] [l.handler.ICloudAccountBridgeHandler] - Starting iCloud authentication (AuthState=INITIAL, Thing=icloud:account:myaccount)...
2025-11-13 22:50:34.341 [DEBUG] [inding.icloud.internal.ICloudService] - Checking session token validity
2025-11-13 22:50:34.343 [DEBUG] [inding.icloud.internal.ICloudSession] - iCloud request POST https://setup.icloud.com/setup/ws/1/validate.
2025-11-13 22:50:34.362 [INFO ] [utomation.script.file.thing_rules.js] - iCloud-Account Status UNKNOWN
2025-11-13 22:50:34.619 [DEBUG] [inding.icloud.internal.ICloudService] - Token is not valid. Attemping new login.
org.openhab.binding.icloud.internal.ICloudApiResponseException: Request https://setup.icloud.com/setup/ws/1/validate failed with 421.
	at org.openhab.binding.icloud.internal.ICloudSession.request(ICloudSession.java:148) ~[?:?]
	at org.openhab.binding.icloud.internal.ICloudSession.post(ICloudSession.java:98) ~[?:?]
	at org.openhab.binding.icloud.internal.ICloudService.validateToken(ICloudService.java:190) ~[?:?]
	at org.openhab.binding.icloud.internal.ICloudService.authenticate(ICloudService.java:94) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.checkLogin(ICloudAccountBridgeHandler.java:338) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.callApiWithRetryAndExceptionHandling(ICloudAccountBridgeHandler.java:164) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:127) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) [?:?]
2025-11-13 22:50:34.623 [DEBUG] [inding.icloud.internal.ICloudService] - Authenticating as notme@icloud.com...
2025-11-13 22:50:34.624 [DEBUG] [inding.icloud.internal.ICloudSession] - iCloud request POST https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true.
2025-11-13 22:50:35.278 [WARN ] [l.handler.ICloudAccountBridgeHandler] - iCloud authentication failed. Invalid credentials.
2025-11-13 22:50:35.283 [INFO ] [utomation.script.file.thing_rules.js] - iCloud-Account Status OFFLINE

```
and traced

```
2025-11-13 23:19:14.979 [TRACE] [inding.icloud.internal.ICloudSession] - Calling https://setup.icloud.com/setup/ws/1/validate
Headers -----
java.net.http.HttpHeaders@db4184c6 { {Origin=[https://www.icloud.com], Referer=[https://www.icloud.com/]} }
Body -----
null
------

2025-11-13 23:19:15.132 [INFO ] [utomation.script.file.thing_rules.js] - iCloud-Account Status UNKNOWN
2025-11-13 23:19:15.252 [TRACE] [inding.icloud.internal.ICloudSession] - Result https://setup.icloud.com/setup/ws/1/validate 421
Headers -----
java.net.http.HttpHeaders@89e06680 { {access-control-allow-credentials=[true], access-control-allow-origin=[https://www.icloud.com], access-control-expose-headers=[X-Apple-Request-UUID,Via], cache-control=[no-cache, no-store, private], connection=[keep-alive], content-length=[1443], content-
type=[application/json; charset=UTF-8], date=[Thu, 13 Nov 2025 22:19:15 GMT], server=[AppleHttpServer/a3fb6e96e80a], strict-transport-security=[max-age=31536000; includeSubDomains;], via=[xrail:icloud-xrail-group109-ext-6d4ddfcc7f-rltdw:8301:25R361:grp109,631194250daa17e24277dea86cf30319:e46
6302997781c14af8f866f129e17c5:deber3], x-apple-edge-response-time=[159], x-apple-request-uuid=[7e9d8fe9-5af2-4a4e-bd79-d73babe5a34f], x-apple-user-partition=[45], x-responding-instance=[setupservice:44500503:pv51p45ic-qukt21011301:8003:2542B423:c588c5ab8ce0]} }
Body -----
{"success":false,"requestInfo":[{"country":"DE","timeZone":"GMT+1","region":"NI"}],"configBag":{"urls":{"accountLoginUI":"https://idmsa.apple.com/appleauth/auth/signin?widgetKey=d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d","accountLogin":"https://setup.icloud.com/setup/w
s/1/accountLogin","accountRepairUI":"https://appleid.apple.com/widget/account/?widgetKey=d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d#!repair","downloadICloudTerms":"https://setup.icloud.com/setup/ws/1/downloadLiteTerms","repairDone":"https://setup.icloud.com/setup/ws/1/r
epairDone","accountCreate":"https://setup.icloud.com/setup/ws/1/createLiteAccount","getICloudTerms":"https://setup.icloud.com/setup/ws/1/getTerms","vettingUrlForPhone":"https://id.apple.com/IDMSEmailVetting/vetSharePhone","accountCreateUI":"https://appleid.apple.com/widget/account/?widgetKey
=d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d#!create","accountVerifyUI":"https://id.apple.com/identity?widgetKey=d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d","accountAuthorizeUI":"https://idmsa.apple.com/appleauth/auth/authorize/signin?client_id=d39b
a9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d","vettingUrlForEmail":"https://id.apple.com/IDMSEmailVetting/vetShareEmail","eventGatewayV1":"https://gateway.icloud.com/rsvp/events/v1"},"accountCreateEnabled":true},"error":"Missing X-APPLE-WEBAUTH-TOKEN cookie"}
------

2025-11-13 23:19:15.254 [DEBUG] [inding.icloud.internal.ICloudService] - Token is not valid. Attemping new login.
org.openhab.binding.icloud.internal.ICloudApiResponseException: Request https://setup.icloud.com/setup/ws/1/validate failed with 421.
        at org.openhab.binding.icloud.internal.ICloudSession.request(ICloudSession.java:148) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudSession.post(ICloudSession.java:98) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudService.validateToken(ICloudService.java:190) ~[?:?]
        at org.openhab.binding.icloud.internal.ICloudService.authenticate(ICloudService.java:94) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.checkLogin(ICloudAccountBridgeHandler.java:338) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.callApiWithRetryAndExceptionHandling(ICloudAccountBridgeHandler.java:164) ~[?:?]
        at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:127) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1583) [?:?]
2025-11-13 23:19:15.274 [DEBUG] [inding.icloud.internal.ICloudService] - Authenticating as notme@icloud.com...
2025-11-13 23:19:15.276 [DEBUG] [inding.icloud.internal.ICloudSession] - iCloud request POST https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true.
2025-11-13 23:19:15.278 [TRACE] [inding.icloud.internal.ICloudSession] - Calling https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true
Headers -----
java.net.http.HttpHeaders@c04f855a { {Accept=[*/*], Content-Type=[application/json], Origin=[https://www.icloud.com], Referer=[https://www.icloud.com/], X-Apple-OAuth-Client-Id=[d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d], X-Apple-OAuth-Client-Type=[firstPartyAuth], X-Apple-OAuth-Redirect-URI=[https://www.icloud.com], X-Apple-OAuth-Require-Grant-Code=[true], X-Apple-OAuth-Response-Mode=[web_message], X-Apple-OAuth-Response-Type=[code], X-Apple-OAuth-State=[auth-2b9b651e-c514-4725-b0ba-5d4af9dc8867], X-Apple-Widget-Key=[d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d]} }
Body -----
{"password":"geheim","accountName":"notme@icloud.com","trustTokens":["HSAFMTKNSRVXWFlalLNWuboDSthEWy5eEwu/dtRoGHjV1Uw+IWJff/LlMY10ah/lXMA+4WNSn5ShsPYvDCDCfsAvitAK8PSBxmHqspHKMIm+gBPvkfuiDzLi3M6Dgqn2rVlLCyhinczZ8uM7VsfGsku9PfXuR4XerkWzlvqTV/mSOPVavTMhJqbn5em69bsD9qNWSRVX"],"rememberMe":true}
------

2025-11-13 23:19:15.752 [TRACE] [inding.icloud.internal.ICloudSession] - Result https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true 503
Headers -----
java.net.http.HttpHeaders@cac3ff8d { {connection=[keep-alive], content-length=[190], content-type=[text/html], date=[Thu, 13 Nov 2025 22:19:15 GMT], server=[Apple]} }
Body -----
<html>
<head><title>503 Service Temporarily Unavailable</title></head>
<body>
<center><h1>503 Service Temporarily Unavailable</h1></center>
<hr><center>Apple</center>
</body>
</html>

------

```

found 1 maybe interesting error in trace log:
`"error":"Missing X-APPLE-WEBAUTH-TOKEN cookie"`

**Comment by f44jwx85y9-ship-it** [comment]
+1, iCloud Binding failed 2/3 days ago and never came back online.

OH 4.3.5 / Binding 4.3.5

**Comment by maihacke** [comment]
It could be related to Apple introducing a new authentication standard (SRP). I started to implement that month ago, but it is not easy without API documentation. Since the "normal" authentication was reactivated by apple a few days later, I stopped that approach. I fear they now enforcing the new authentication again, until it is implemented or apple changes his mind again, the binding won't work. So somebody has to adopt the existing implementation from other projects for the icloud binding.

See also https://medium.com/@vortj/issue-fixing-the-apple-id-auth-issue-in-fastlane-sharing-what-worked-for-me-today-97ca431ed0a5

**Comment by mike-bike** [comment]
Thanks for the pointer @maihacke. I think I understand the cause, but I don't have the knowledge to fix it. I then need to be patient and wait for a more knowledgeable _somebody_ to implement the solution.

Have a good weekend!

**Comment by cgatnet** [comment]
@kaikreuzer maybe you can help with your experience and overview of openhab, if there is a problem with the .jar?

**Comment by maihacke** [comment]
What one need to implement this is a good view on the binding (which is not complex to learn) and understanding of the new apple authentication theme. I implemented the now outdated authentication along an existing python project. they currently also face problems, but it seems there is a PR which fixes that (https://github.com/picklepete/pyicloud/pull/459#issuecomment-3544551510).
As already that a started implementing this month ago, but stopped, since the existing auth scheme worked again.
So I see two options:

1. implement SRP simliar to to pyicloud in the existing binding
2. create new binding leveraging pyicloud directly

**Comment by bredmich** [comment]
Hey Simon,

any news?

Cheers

**Comment by maihacke** [comment]
I started to implement the new auth scheme. It consists of two steps. One is working. The second more complicated one not. So it takes more time.

**Comment by lochmueller** [comment]
@maihacke Is there anything I can do to help? Either by testing a fresh binding, or perhaps you have an Amazon wish list, and I could fulfill a Christmas wish (send me a DM with your wishlist link)? :-) Thank you for looking into this...

**Comment by maihacke** [comment]
❤️ Thank you very much. Currently I can't see a way to support, I need to compare the exact computation pyicloud does with my rewrite. Problem is that Apple does not use the normal SRP standard, so existing libraries for that do not work, and the Cloud API doesn't give meaningful errormessages.

Thanks a lot for offering a christmas gift, but since I did not deliver a solution yet I fell that would be to much ;-)

**Comment by maihacke** [comment]
I managed to create a first working (=works for me) implementation. A published a release candidate here: https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-5.0.x-beta-0

Please test and give feedback

**Comment by ErikDB87** [comment]
Jar installed, and both the bridge as the iphone things are online!

Alas, after reboot (and a later test confirms: after an openHAB restart), the bridge requires the 2FA anew...

**Comment by bredmich** [comment]
I´m back online, works as expected.
Thanks for your work Simon!

**Comment by ErikDB87** [comment]
> I´m back online, works as expected.
> Thanks for your work Simon!

Did you try restarting openHAB?

**Comment by bredmich** [comment]
> Did you try restarting openHAB? 

I'll let someone else try that first :-)

**Comment by cgatnet** [comment]
I installed the snapshot and got this: 
Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 401.
... digging... :)

**Comment by ErikDB87** [comment]
I assume the authentication token isn't stored or loaded properly? But that's mere speculation.

**Comment by cgatnet** [comment]
> I installed the snapshot and got this: Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 401. ... digging... :)

OK, I used an "app specific password", because I was digging into the last error.

Now switched back to my "normal" password and it went ONLINE!
WOHOOOO!!!! 👍

**Comment by cgatnet** [comment]
> > I installed the snapshot and got this: Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 401. ... digging... :)
> 
> OK, I used an "app specific password", because I was digging into the last error.
> 
> Now switched back to my "normal" password and it went ONLINE! WOHOOOO!!!! 👍

Did a normal restart of the openhab service on my Raspberry Pi 4 and it ran into a "invalid credentials".
But i just missed to remove the original icloud binding from the addon.cfg file before restarting.
Fixed that and brought it ONLINE again with that SNAPSHOT version and did a new restart from here.

It came up with the request for the 2FA code, as reported by ErikDB87 and went ONLINE with it again! 
WOHOOO! :)

**Comment by ErikDB87** [comment]
> It came up with the request for the 2FA code, as reported by ErikDB87 and went ONLINE with it again!

I assume you meant after manually entering the 2FA code again, right?

**Comment by cgatnet** [comment]
> I managed to create a first working (=works for me) implementation. A published a release candidate here: https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-5.0.x-beta-0
> 
> Please test and give feedback

Thank you so much for your time! I will test if you need another test.

> > It came up with the request for the 2FA code, as reported by ErikDB87 and went ONLINE with it again!
> 
> I assume you meant after manually entering the 2FA code again, right?

yes, exactly!

**Comment by mike-bike** [comment]
This sounds promising.

@maihacke, firstly many thanks from me as well.

Unfortunately, I cannot get around the 401

I am running OpenHABian on a PI5 with openHAB 5.0.3. Any suggestions?

Thanks,
Michael

**Comment by cgatnet** [comment]
> This sounds promising.
> 
> [@maihacke](https://github.com/maihacke), firstly many thanks from me as well.
> 
> Unfortunately, I cannot get around the 401
> 
> I am running OpenHABian on a PI5 with openHAB 5.0.3. Any suggestions?
> 
> Thanks, Michael

First idea...
... make sure your openhab is using the snapshot version you downloaded by looking into UI addon page or karaf console.

**Comment by mike-bike** [comment]
I think it does. However, openHAB came up with the delivered version.

```
openhab> bundle:list | grep -i icloud
350 │ Active │  80 │ 5.0.4.202512041635    │ openHAB Add-ons :: Bundles :: iCloud Binding
```

 I will try another restart.

**Comment by mike-bike** [comment]
Hmm, restart with cache cleared did not help. Might be something weird with my system. 

```
CONFIGURATION_ERROR
Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 401.
```

I do see below messages in the log which look suspicous to me uneducated user...

```
2025-12-04 20:44:55.977 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Locating method activate in class org.openhab.binding.icloud.intern
al.ICloudHandlerFactory
2025-12-04 20:44:55.981 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Declared Method org.openhab.binding.icloud.internal.ICloudHandlerFa
ctory.activate([interface org.osgi.service.component.ComponentContext]) not found
2025-12-04 20:44:55.983 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Locating method activate in class org.openhab.core.thing.binding.Ba
seThingHandlerFactory
2025-12-04 20:44:55.984 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Found activate method: protected void org.openhab.core.thing.bindin
g.BaseThingHandlerFactory.activate(org.osgi.service.component.ComponentContext)
2025-12-04 20:44:55.986 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : invoking activate: activate: parameters [org.apache.felix.scr.impl.
manager.ComponentContextImpl]
2025-12-04 20:44:55.989 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : invoked activate: activate
2025-12-04 20:44:55.991 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Set implementation object for component
2025-12-04 20:44:55.992 [DEBUG] [icloud.internal.ICloudHandlerFactory] - bundle org.openhab.binding.icloud:5.0.4.202512041635 (305)[
org.openhab.binding.icloud.internal.ICloudHandlerFactory(394)] : Changed state from satisfied to active
```

**Comment by mike-bike** [comment]
Ok, Guys! Seems I have managed to get it working. My openHAB 5.0.3 seemed to be unstable - I had it downgraded from 5.1.0M2 for the iCloud binding test. After several unsuccessful restarts, I have then loaded current milestone 5.1.0M3 and retried the new binding. The many attempts annoyed Apple and they had locked my account. I don't know why this has happened, because the password provided was definitely correct. 
However, the new binding returned the correct error code 403 and the log states `Account Locked`. I followed the instructions to unlock my account and used the opportunity to change my password. Second try with the new password triggered finally the 2FA process. After entering the code into openHAB, the thing came online. 

**I can now confirm that the new binding is working with openHAB 5.1.0M3.**

@maihacke, many thanks for your efforts, dedication and commitment to fix the authentication issue.

Regards,
Michael

**Comment by bredmich** [comment]
Could it be that the discovery doesn´t work anymore?
I can´t provide debug logs as nothing happens, when pressing the scan button :-(

**Comment by ErikDB87** [comment]
> Could it be that the discovery doesn´t work anymore?
> I can´t provide debug logs as nothing happens, when pressing the scan button :-(

Maybe after you clear the cache?

**Comment by bredmich** [comment]
> Maybe after you clear the cache?

Yes, that worked.
But it´s strange that there´s no log message for the scan.

**Comment by lochmueller** [comment]
Hmm... I got only `Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 403.` (Note. 403. Note 401 like @cgatnet above). I try to get more details....

The `iCloud request POST https://idmsa.apple.com/appleauth/auth/signin/init.`in front of the 403 got a 200 status code. and a content that looks like some JSON incl. tokens.

**Comment by mike-bike** [comment]
403 means the Account is locked. If you set binding to trace you can see the error message in the log . Happened the same to me. You need to log on to icloud.com to reset. Be prepared to chose a new password. Good luck. GrußMichaelAm 07.12.2025 um 20:39 schrieb Tim ***@***.***>:﻿lochmueller left a comment (openhab/openhab-addons#17590)
Hmm... I got only Request https://idmsa.apple.com/appleauth/auth/signin/complete?isRememberMeEnabled=true failed with 403. (Note. 403. Note 401 like @cgatnet above). I try to get more details....

**Comment by lochmueller** [comment]
@mike-bike thanks. You are right... I double-checked it today and check it on icloud.com. But I think the account was locked again after my first tests, and I had still a valid icloud.com session on icloud.com. So note for every out there. **Logout** & Login on icloud.com to check if the account is locked 👍

**Comment by ErikDB87** [comment]
@maihacke, I don’t know whether you’ve had the chance to look at the ‘after OH restart’ problem?

**Comment by maihacke** [comment]
No not yet. Have to find time.

**Comment by ErikDB87** [comment]
I wish I could help, but
1. my technical capabilities are rather low - but that wouldn't stop me from at least trying, but
2. I don't have anything Apple... It's my girlfriend who's got the iPhone I want to include in openHAB. And I can't constantly ask her to clear the 2FA... Unless there's a way around that, with an emulated iPhone and an account of my own?

**Comment by ccutrer** [comment]
@maihacke: is there a particular reason you haven't submitted a PR yet for your fix so far? I understand that there are some cases that still might be problematic, but it seems like it is working for many people, and we're getting very close to the 5.1 release.

**Comment by mike-bike** [comment]
I second that comment. Latest fixed iCloud binding is working for me. I’d appreciate to get it into 5.1, unless the author is having severe concerns. But the provided version does not work at all, so why not?

Thanks for all your efforts.

**Comment by maihacke** [comment]
The reason is that the code quality wouldn’t meet the requirements, eg missing javadoc.

**Comment by ccutrer** [comment]
Are you hoping to have time this week to address that? Would you be okay if someone else did that for you (not necessarily volunteering myself, but maybe; I've just been watching this issue and hoping it would make it into 5.1. The icloud binding isn't critical for me)?

**Comment by ErikDB87** [comment]
> Would you be okay if someone else did that for you (not necessarily volunteering myself, but maybe)?

I can't speak for @maihacke, of course, but isn't that a bit the idea of open source code? 😇

**Comment by ccutrer** [comment]
😂 You're not wrong. But it's not like @maihacke has abandoned anything for a long time and it needs picked up. And it can be frustrating to some people if you pick up their work while they're still in the middle of it, and thus discourage future contributions. Not to mention that if he's actually working on it without me checking, and then I do it, it might be a duplication of effort, when it might be more valuable for the community for me to spend my limited hobby time on something else.

**Comment by falseat** [comment]
# openHAB iCloud Binding (BETA) – Summary of Troubleshooting Attempts

This document summarizes all troubleshooting steps and observations while trying to get the **BETA version of the new iCloud binding** working in openHAB. Unfortunately, authentication has not succeeded so far.

The goal of this summary is to give a concise, technical overview of what has already been tried and what is currently observed.

---

## Environment / Setup

- Official iCloud binding **uninstalled**
- Only the **BETA iCloud binding JAR** is present (no parallel instances)
- Bridge deleted
- New bridge created multiple times using:
  - `.things` file (text-based)
  - openHAB UI
- openHAB restarted after most changes - including 'clean-cache'

---

## Observed Authentication Errors

Authentication fails consistently during login with various HTTP response codes:

- **401 Unauthorized**
- **403 Forbidden**
- **503 Service Unavailable**



### Common Pattern
- The **first authentication-related request often returns HTTP 200**
- A **subsequent request fails** with 401 / 403 / 503
- The **2FA challenge is never triggered**, unlike earlier behavior (this worked until around mid-November)

---

## Cleanup & Reset Attempts

All of the following steps were tried multiple times:

- Deleted the iCloud bridge and recreated it
- Removed `.thing` files
- Deleted all `icloud*.json` files (including backups) from: /var/lib/openhab/jsondb
- Restarted openHAB after each change
- Ran: 'openhab-cli clean-cache'
- Deleted and reinstalled the BETA `.jar`
- Logged into `icloud.com` and logged out of **all iCloud browser sessions**
- Tried to created an app-specific password and testet with this

**Result:**
- A new `icloud.json` file is always recreated as expected
- Authentication still fails with the same HTTP errors

---

## Additional Debug Observation

While reviewing debug logs, I noticed a potential inconsistency:

- The `scnt` value in the **session data** does **not match** the value stored in the corresponding `icloud*.json` file
- The value appears to match the **first successful (HTTP 200) request**
- It does **not match** the second request, which then fails with **401 Unauthorized**

I am **not a developer**, so I cannot determine whether this is causal or merely a symptom.

---

## Preliminary Analysis (Logs + Code Review)

By correlating debug logs with the source code (`ICloudService.java`), I arrived at the following *tentative* conclusions:

- The `-20101` error does **not** indicate:
- wrong credentials
- a network connectivity issue
- Authentication appears to fail during the **SRP login phase**, *before* 2FA is initiated

### Potential Root Cause
According to Copilot, Apples authentication flow changes that are no longer compatible with the current implementation of the binding.

Possible contributing factors:
- Use of an outdated or hard-coded Apple client ID
- Missing browser-like headers (e.g. User-Agent, language, fingerprint data)

This could explain:
- Why credentials seem to be accepted initially
- Why Apple rejects the login context before 2FA
- Why different HTTP error codes (401 / 403 / 503) are observed

---

## Open Questions

- Does this assessment align with your understanding, or is it incorrect?
- Is there anything specific I can test, log, or instrument to help debug this further?
- Is there an update of the initial beta *.jar?

I am happy to assist in any way (testing, logging, experiments).

**Comment by ccutrer** [comment]
I went ahead and prepped a PR :). I rebased the changes onto `main` (@maihacke had developed them against the 5.0 branch), and then did some code cleanup and filled out some javadocs.

**Comment by maihacke** [comment]
I appreciate very much your effort to help here. Sadly part of the changes overlap with changes I started in my local development (javadoc) or do not go in the direction I planed (e.g. the cluttered helper methods in SrpAuthentication, which need to be cleaned up and are result of my experiments. On top there are some security checks missing in SRP.. No offense, but no it would mean additional work for me to understand the changes you made and merge them...
I'm not sure if this is the best way to go forward.

**Comment by maihacke** [comment]
I had a short look in to the restart problem and found the solution.

**Comment by maihacke** [comment]
I created a new release for the fix https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-5.0.x-beta-1.

PR for upcoming release is here: https://github.com/openhab/openhab-addons/pull/19832

@ccutrer If you see an further improvements please use this PR, it was just easier for me to start from my local development for fixing the issue.

**Comment by FranzSchi** [comment]
@maihacke ,
thanks for the fix. if i try it it in OH 5.0.3 get this message:

```
openhab> bundle:list | grep -i cloud
672 x Installed x  80 x 5.0.4.202512191349    x openHAB Add-ons :: Bundles :: iCloud Binding
721 x Active    x  80 x 5.0.3                 x openHAB Add-ons :: Bundles :: IO :: openHAB Cloud Connector
openhab> bundle:start 672
Error executing command: Error executing command on bundles:
        Error starting bundle 672: Could not resolve module: org.openhab.binding.icloud [672]
  Unresolved requirement: Import-Package: org.bouncycastle.crypto; version="[1.81.0,2.0.0)"

```

**Comment by bredmich** [comment]
@maihacke same error als FranzSchi:
```org.osgi.framework.BundleException: Could not resolve module: org.openhab.binding.icloud [363]
  Unresolved requirement: Import-Package: org.bouncycastle.crypto; version="[1.81.0,2.0.0)"

	at org.eclipse.osgi.container.Module.start(Module.java:463)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:445)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundle(DirectoryWatcher.java:1260)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.startBundles(DirectoryWatcher.java:1233)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:520)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:365)
	at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:316)```

**Comment by cwihne** [comment]
I tested both beta-0 and new beta-1 of today, both worked. I had doubts, as jar filesize decreased from 11MB to 70KB. Maybe relevant for the above reports?

**Comment by maihacke** [comment]
Yes something went wrong with the local maven build. Just checking. The error occurs only if you restart openhab and not just replace the jars. Thats why I didn't notice it first time.

**Comment by bredmich** [comment]
> The error occurs only if you restart openhab and not just replace the jars. Thats why I didn't notice it first time.

I replaced the JAR file and got the error.

**Comment by maihacke** [comment]
I uploaded a new jar, it now 11MB again, thanks for the hint.

https://github.com/maihacke/openhab-addons/releases/download/17590-srp-5.0.x-beta-1/org.openhab.binding.icloud-5.0.4-SNAPSHOT.jar

**Comment by FranzSchi** [comment]
The dependency problem seems to be resolved now.

**Comment by ccutrer** [comment]
> I appreciate very much your effort to help here. Sadly part of the changes overlap with changes I started in my local development (javadoc) or do not go in the direction I planed (e.g. the cluttered helper methods in SrpAuthentication, which need to be cleaned up and are result of my experiments. On top there are some security checks missing in SRP.. No offense, but no it would mean additional work for me to understand the changes you made and merge them... I'm not sure if this is the best way to go forward.

Not a problem. I'm not offended at all. I'm glad you found some more time to work on it. I can confirm that I no longer have to re-2FA every time my bridge thing initializes with your new PR! Good work!

**Comment by falseat** [comment]
Unfortunately I still have the same issues with error 401 and I have absolutly no clue what I could do in order to make it working again :(

**Comment by ErikDB87** [comment]
> I uploaded a new jar, it now 11MB again, thanks for the hint.
> 
> https://github.com/maihacke/openhab-addons/releases/download/17590-srp-5.0.x-beta-1/org.openhab.binding.icloud-5.0.4-SNAPSHOT.jar

It works! 🎉

**Comment by maihacke** [comment]
I created two new test versions. The versions should have not functional changes, but address codestyle issue which had to be fixed for the upcoming OH release.

One version is for the current stable 5.0.x versions:
https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-5.0.x-beta-3

For the upcoming 5.1 release the version is here:
https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-main-beta-0

Since I have no 5.1. installation the second version is untested by me. So it would be nice if somebody has a beta version running to tests this.

**Comment by mike-bike** [comment]
> I created two new test versions. The versions should have not functional changes, but address codestyle issue which had to be fixed for the upcoming OH release.
> 
> One version is for the current stable 5.0.x versions: https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-5.0.x-beta-3
> 
> For the upcoming 5.1 release the version is here: https://github.com/maihacke/openhab-addons/releases/tag/17590-srp-main-beta-0
> 
> Since I have no 5.1. installation the second version is untested by me. So it would be nice if somebody has a beta version running to tests this.

Hi @maihacke, many thanks for your continous effort. I have upgraded the 5.1.0 version over the previous 5.0.4 version on a running openHAB 5.1.0.RC1 instance. The iCloud Account thing switched to `UNINITIALIZED` and then quickly to `ONLINE` again. It didn't even require another 2FA. Other thing devices quickly returned to green as well. All items seem to receive updates as expected. I will monitor the situation over the next few days and may consider loading RC3 which requires a restart then.
I will keep you posted.

Thanks again for your valuable support.

I am wishing you a Merry Christmas season and a Happy New Year 
Michael

**Comment by mike-bike** [comment]
**Update:** I couldn't resist and just upgraded to fresh openHAB 5.1.0 release. It took a few restarts to get rid of rrd4j persistence load issues (nothing todo with iCloud binding I assume). It seems that current GA still includes a prev version of iCloud. The binding did not come on:
```
openhab> bundle:list | grep -i icloud
303 │ Active │  80 │ 5.1.0.RC1             │ openHAB Add-ons :: Bundles :: iCloud Binding
```

After manually updating with your beta binding
```
openhab> bundle:update 303 file:///home/openhabian/addons/org.openhab.binding.icloud-5.1.0-SNAPSHOT.jar
openhab> bundle:list | grep -i icloud
303 │ Active │  80 │ 5.1.0.202512221036    │ openHAB Add-ons :: Bundles :: iCloud Binding
```
all iCloud things came up `ONLINE`.

iCloud seems to be running fine😀

**Comment by bredmich** [comment]
> One version is for the current stable 5.0.x versions:

Just replaced the jar and iCloud is back online and running.
Thanks!

**Comment by cgatnet** [comment]
Also tried 5.0.4 in new Version. Works for me. 
I updated while I was in a 2FA authentication and after the update the 2FA was no longer required :)
